must-call-checker/README.md

```diff
@@ -0,0 +1,88 @@
+and reports cases where required methods might not have been called.
+
+
+### Explanation of qualifiers and type hierarchy
+
+The MustCall Checker supports two qualifiers: `@MustCall` and `@MustCallTop`. The `@MustCall`
+qualifier's arguments are the methods that the annotated type
+might be responsible for calling. The type `@MustCall({"a"}) Object obj` means
+that at run time `obj` cannot have an obligation to call any method other than `a()` (but it may or 
+may not actually have an obligation to call `a()`).
+Likewise, `@MustCall({"m1", "m2"}) Object` represents all objects that need to
+call `m1`, `m2`, both, or neither. It should not represent an object that needs
+to call some *other* method.  The type hierarchy is:
+
+                     @MustCallTop
+                           |
+                          ...
+                           |
+               @MustCall({"m1", "m2"})     ...
+                    /             \
+       @MustCall({"m1"})     @MustCall({"m2"})  ...
+                    \            /
+                    @MustCall({})
+
+Note that `@MustCall({"m1", "m2"}) Object` represents more possible values than
+`@MustCall({"m1"}) Object` does, because `@MustCall({"m1"}) Object` can only
+contain an object that needs to call `m1` or one that needs to call nothing - an
+object that needs to call `m2` (or both `m1` and `m2`) is not permitted.
+
+The default type in unannotated code is `@MustCall({})`.  Given an expression of
+unannotated type, its value definitely has no obligations to call any methods.
+
+This default is unsound in the sense that a priori the checker
+has no knowledge of what methods an object might need to call. A sound default would be
+`@MustCallTop`, which represents all objects, regardless of what methods they might have an obligation to call.
+Such a default, however, would not be **useful**, because it is desirable to track objects that have obligations
+to call specific methods, not all objects generally.
+The checker therefore uses the unsound default `@MustCall({})` for unannotated types,
+but permits a user to annotate a type with another `@MustCall` type to express that a particular type has
+a particular obligation. Such an annotation is then tracked soundly. 
+
+
+### Example use
+
+If a user annotates a class declaration with a `@MustCall(...)` annotation, any use of the class's
+type has that `@MustCall` annotation by default. For example, given the following class annotation:
+
+    package java.net;
+    
+    import org.checkerframework.checker.mustcall.qual.MustCall;
+    
+    @MustCall({"close"}) class Socket { }
+    
+any use of `Socket` defaults to `@MustCall({"close"}) Socket`.
+
+TODO: Move the following implementation comment out of the user documentation.
+TODO: Make the `@MustCall` annotation inherited.
+
+
+### Implementation details and caveats
+
+This type system cannot be used to track a must-call obligation for a `String`, because it treats all
```

Comment:
The second part of this sentence does not imply the first part.  This type system *can* track must-call obligations for strings.  Also, all String concatenations happen to have the type `@MustCall({})`.

I'm not sure the latter fact is worth calling out explicitly.  All newly-created strings have type `@MustCall({})`, and string concatenations are one example of that.

Is the problem that the AST lacks a call to `toString`?  (That is, if the toString() were explicit there would be no problem?)  The CFG contains an explicit `toString` call, so maybe you can solve your problem in the transfer function rather than by special-casing string concatenation.

Reply:
Maybe I should have replaced the first part of the sentence with "This type system ~~cannot~~ should not be used to track a must-call obligation for a `String`", because the special treatment of string concatenation would make doing so unsound (e.g. the result of concatenating two strings with types `@MustCall({"a"}) String` and `@MustCall({"b"}) String` would be...`@MustCall({}) String`, because all string concatenations have that type. I think that this is at least worth mentioning, since it might violate a user's expectations.

> All newly-created strings have type @MustCall({}), and string concatenations are one example of that.

It goes beyond that, which is why this is mentioned. For example, if the user for some reason used the following stub file:
```
package java.lang;
import org.checkerframework.checker.mustcall.qual.MustCall;
@MustCall("matches") class String { }
```
the result of every string concatenation would still be `@MustCall({}) String`, even though newly-created strings would have the type `@MustCall({"matches"}) String`.

> Is the problem that the AST lacks a call to toString? (That is, if the toString() were explicit there would be no problem?) The CFG contains an explicit toString call, so maybe you can solve your problem in the transfer function rather than by special-casing string concatenation.

Yes, that is the problem. I tested that explicitly when I made this fix (see the file `ToStringOnSocket.java` in the mustcall tests). Fixing it in the transfer function when encountering a `toString` call is a good idea that I didn't consider; I'll look into it.